File: narrative_memory/story_ledger.py
import json
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Dict, Optional


class StoryLedger:
    """Ledger that stores narrative actions as stories."""

    def __init__(self, filename: str = "narrative_ledger.jsonl"):
        self.filepath = Path(filename)
        # Ensure the file exists
        self.filepath.touch(exist_ok=True)

    def log_action(self, actor: str, action: str, context: Optional[Dict[str, Any]] = None) -> None:
        """Log a narrative action by appending it as JSON."""
        entry = {
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "actor": actor,
            "action": action,
        }
        context = context or {}
        for key in ("traceId", "reqId", "ctrId", "caseId", "eid", "status"):
            value = context.get(key)
            if value:
                entry[key] = value

        for key in ("vids", "zkRefs", "evidenceRefs"):
            value = context.get(key)
            if isinstance(value, list):
                normalized = sorted({str(item).strip() for item in value if str(item).strip()})
                if normalized:
                    entry[key] = normalized

        with self.filepath.open("a", encoding="utf-8") as f:
            f.write(json.dumps(entry, ensure_ascii=False) + "\n")

    def stories(self):
        """Yield stories from the ledger."""
        with self.filepath.open("r", encoding="utf-8") as f:
            for line in f:
                if line.strip():
                    yield json.loads(line)


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description="Narrative memory ledger")
    parser.add_argument("actor", help="Nombre del actor")
    parser.add_argument("action", help="Descripción de la acción")
    parser.add_argument(
        "--file",
        default="narrative_ledger.jsonl",
        help="Archivo del libro mayor para almacenar historias",
    )
    args = parser.parse_args()

    ledger = StoryLedger(args.file)
    ledger.log_action(args.actor, args.action)
    print(f"Historia agregada al libro mayor {args.file}")

File: backend/api/storyLedger.ts
import fs from "fs";
import path from "path";

export type StoryLedgerContext = {
  caseId?: string;
  traceId?: string;
  reqId?: string;
  ctrId?: string;
  eid?: string;
  vids?: string[];
  zkRefs?: string[];
  evidenceRefs?: string[];
  status?: string;
};

function normalizeStringArray(values?: string[]): string[] | undefined {
  if (!Array.isArray(values)) {
    return undefined;
  }
  const normalized = [...new Set(values.map((value) => value.trim()).filter(Boolean))].sort();
  return normalized.length > 0 ? normalized : undefined;
}

export class StoryLedger {
  private filepath: string;

  constructor(
    filename?: string
  ) {
    this.filepath = filename ?? "";
    if (this.filepath) {
      this.ensureFilepath(this.filepath);
    }
  }

  private resolvePath(): string {
    const resolved = process.env.CATALYST_NARRATIVE_LEDGER_PATH ?? this.filepath;
    if (!resolved) {
      return path.join(process.cwd(), "narrative_memory", "narrative_ledger.jsonl");
    }
    return resolved;
  }

  private ensureFilepath(target: string): void {
    const dir = path.dirname(target);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
    if (!fs.existsSync(target)) {
      fs.writeFileSync(target, "");
    }
  }

  logAction(actor: string, action: string, context: StoryLedgerContext = {}) {
    const targetPath = this.resolvePath();
    this.ensureFilepath(targetPath);

    const entry: Record<string, unknown> = {
      timestamp: new Date().toISOString(),
      actor,
      action,
    };

    if (context.traceId) entry.traceId = context.traceId;
    if (context.reqId) entry.reqId = context.reqId;
    if (context.ctrId) entry.ctrId = context.ctrId;
    if (context.caseId) entry.caseId = context.caseId;
    if (context.eid) entry.eid = context.eid;
    if (context.status) entry.status = context.status;

    const vids = normalizeStringArray(context.vids);
    const zkRefs = normalizeStringArray(context.zkRefs);
    const evidenceRefs = normalizeStringArray(context.evidenceRefs);
    if (vids) entry.vids = vids;
    if (zkRefs) entry.zkRefs = zkRefs;
    if (evidenceRefs) entry.evidenceRefs = evidenceRefs;

    fs.appendFileSync(targetPath, JSON.stringify(entry) + "\n", { encoding: "utf8" });
  }
}
